### PR TITLE
CI: Rework how docs are built

### LINKED ISSUE #10555: Doc tests should not take long
It looks like running documentation tests in CI is pretty slow right now. I don't have an exact number yet, but it seems much longer than 10m. I think it should only take a few minutes to tests doc changes, I'd also consider auto-merge on approval.

### PR DESCRIPTION
**Pull request by errordeveloper**
The aim is to make feedback on documentation PR faster, this probably the first phase of changes, we might want to add preview site deployments etc.

Fixes #10555

This runs very fast (under a minute) on GitHub Actions: https://github.com/errordeveloper/cilium/runs/505646021?check_suite_focus=true

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/10572)

### CODE REVIEW

File: .github/workflows/docs.yaml
Comment (errordeveloper) [inline .github/workflows/docs.yaml]: I pushed my own image for now, let's discuss where this one should live and if we need to tag and push it automatically.

**Comment by errordeveloper** [comment]
test-docs-please

**Comment by errordeveloper** [comment]
(Just checking if the old way still works).

**Comment by errordeveloper** [comment]
test-docs-please

File: jenkinsfiles/docs.Jenkinsfile
Comment (errordeveloper) [inline jenkinsfiles/docs.Jenkinsfile:47]: Should we actually keep this job or disable it to reduce load Jenkins (if nothing else)?

**Comment by errordeveloper** [comment]
Actually, what's a little confusing is that there two jobs - `Cilium-Doc-Test` and `Cilium-PR-Doc-Tests`, click on links in the status here leads to the same page...

File: Documentation/check-build.sh
Comment (errordeveloper) [inline Documentation/check-build.sh:75]: I attempted fixing some of the broken links (https://github.com/cilium/cilium/pull/10576), but it looks like Sphinx' link checker will ways fail on some of our links (e.g. things it spots in our docs like `curl https://exampe.com/$(uname)`), it also seems to fail to find anchors also.

Perhaps we should just drop this, or explore ways of controlling the behaviour?

File: jenkinsfiles/docs.Jenkinsfile
Comment (joestringer) [reply jenkinsfiles/docs.Jenkinsfile:47]: I'm all for disabling if this is handled by another CI target and that target is working as intended.

File: .github/workflows/docs.yaml
Comment (joestringer) [reply .github/workflows/docs.yaml]: We already have https://hub.docker.com/r/cilium/docs-builder, presumably we will just push a new version of that image?

File: Documentation/check-build.sh
Comment (joestringer) [reply Documentation/check-build.sh:75]: Just because we don't run the target doesn't mean links stop being dead ;)

Since we don't regularly run this check, some links have rotten. It'd be nice to be able to automate this as well, but I wouldn't make fixing this target a precondition to merging this PR.

File: Documentation/check-build.sh
Comment (joestringer) [inline Documentation/check-build.sh]: nit:
```suggestion
        describe_spelling_errors
        hint_about_wordlist_update
	status_ok=1
    fi
```

(I think your editor introduced some tabs in here unintentionally)

File: Documentation/check-build.sh
Comment (joestringer) [inline Documentation/check-build.sh:56]: I *think* that previously if the spellchecker build failed (non-zero exit code), we would exit immediately. With this script, I suspect we will just skip this conditional block and proceed to attempt to build again without the spellchecker target, so you'll end up running sphinx twice in such cases?

File: Documentation/check-build.sh
Comment (joestringer) [inline Documentation/check-build.sh:32]: I think that the `grep` match here is not quite having the intended effect, because the return code is based on the matched expressions (even if the output for those matches is excluded), it's not based on whether there is other content in the file.

Maybe for the first run, you could just run this grep then pipe the output to `wc -l` and if there's more than zero lines, we need to print them out and fail out.

See the ignored WARNING below about the `admin_guide` link, the build was skipping past that and only complaining about my spelling:

```
$ git diff                                                                                                                                                                     [0/1878]
diff --git a/Documentation/troubleshooting.rst b/Documentation/troubleshooting.rst
index f2e1d29a9947..4f6c896f068f 100644
--- a/Documentation/troubleshooting.rst
+++ b/Documentation/troubleshooting.rst
@@ -6,6 +6,8 @@

 .. _admin_guide:

+@Im'a shootin' for trouble
+
 ###############
 Troubleshooting
 ###############
$ make -C Documentation/
Sending build context to Docker daemon  10.24kB
Step 1/5 : FROM docker.io/library/python:3.7.5-alpine3.10
 ---> 8922d588eec6
Step 2/5 : LABEL maintainer="maintainer@cilium.io"
 ---> Using cache
 ---> be4f82463a09
Step 3/5 : RUN apk add --no-cache --virtual --update     aspell-en     bash     ca-certificates     enchant     git     py-pip     python     sphinx-python     && true
 ---> Using cache
 ---> 3d659eafb699
Step 4/5 : ADD ./requirements.txt /tmp/requirements.txt
 ---> Using cache
 ---> ef81ad334356
Step 5/5 : RUN pip install -r /tmp/requirements.txt
 ---> Using cache
 ---> 17243537fa83
Successfully built 17243537fa83
Successfully tagged cilium/docs-builder:latest
  GEN   Documentation/_api
  GEN   Documentation/_build/html
Validating documentation (syntax, spelling)...
/usr/local/lib/python3.7/site-packages/sphinx/application.py:402: RemovedInSphinx20Warning: app.info() is now deprecated. Use sphinx.util.logging instead.
  RemovedInSphinx20Warning)
/src/Documentation/index.rst:32: WARNING: undefined label: admin_guide (if the link has no caption the label must precede a section header)
/usr/local/lib/python3.7/site-packages/sphinx/application.py:388: RemovedInSphinx20Warning: app.warning() is now deprecated. Use sphinx.util.logging instead.
  RemovedInSphinx20Warning)
WARNING: Found 2 misspelled words
WARNING: Not copying tabs assets! Not compatible with spelling builder

Please fix the following spelling mistakes:
* Documentation/troubleshooting.rst:9: (Im'a)
* Documentation/troubleshooting.rst:9: (shootin)

If the words are not misspelled, run:
Documentation/update-spelling_wordlist.sh Im\\'a shootin
make: *** [Makefile:28: html] Error 1

```

File: Documentation/Makefile
Comment (joestringer) [inline Documentation/Makefile:42]: Can we make this run as a regular user UID/GID?

I ran the following commands and hit trouble:

```
$ make -C Documentation/
...
$ make -C Documentation/ clean
rm: cannot remove '_build/doctrees/environment.pickle': Permission denied
rm: cannot remove '_build/doctrees/gettingstarted/k8s-install-connectivity-test.doctree': Permission denied
rm: cannot remove '_build/doctrees/gettingstarted/kube-router.doctree': Permission denied
rm: cannot remove '_build/doctrees/gettingstarted/k8s-install-azure-cni-steps.doctree': Permission denied
...
$ ls -lh Documentation/_build/doctrees/environment.pickle
-rw-r--r-- 1 root root 704K Mar 13 10:17 Documentation/_build/doctrees/environment.pickle
```

**Review by joestringer — CHANGES_REQUESTED** [review]
This is looking much simpler, awesome!

I think that there's some basic behavioural differences still between the existing implementation and the new one, and ideally I don't want to reintroduce regressions where the warnings are ignored because we've been there before and things get out of hand quickly ;) I did a little bit of investigation below, hopefully that helps guide getting the new version right here.

Unfortunately I don't think we have any sort of basic testing bootstrap; what I'd ideally want is some very simple patches that deliberately add different types of errors into the docs, then run the docs targets to make sure they fail --- I'm thinking syntax errors, spelling errors, cmdref differences.

On cmdref differences, I didn't investigate this, but I'd want to make sure that if you add a flag to the daemon, the `cmdref` target here correctly generates the new version of the docs, and the docs test build will fail because the developer hasn't added those cmdref changes to the documentation.

File: .github/workflows/docs.yaml
Comment (errordeveloper) [reply .github/workflows/docs.yaml]: For sure, I just need push access to begin with :)

I'd be happy to automate pushing the image, but perhaps best to do it in another PR?

File: Documentation/Makefile
Comment (errordeveloper) [reply Documentation/Makefile:42]: Good point, I'll see what can be done. I only tested on Docker for Mac, and the [`osxfs`](https://docs.docker.com/docker-for-mac/osxfs/) does remaps UID/GID.

File: Documentation/Makefile
Comment (errordeveloper) [inline Documentation/Makefile:17]: @joestringer this should work fine on Linux, would you mind to test please?

File: Documentation/Makefile
Comment (errordeveloper) [reply Documentation/Makefile:42]: See https://github.com/cilium/cilium/pull/10572/files#r392967432.

File: Documentation/check-build.sh
Comment (errordeveloper) [reply Documentation/check-build.sh:75]: Sure, we should keep up with links, but we have some links that will keep on failing because of how this link checker works (see https://github.com/cilium/cilium/pull/10576#issuecomment-599504742).

File: Documentation/check-build.sh
Comment (errordeveloper) [reply Documentation/check-build.sh]: yeah... I tend to mix vim/vscode for shell scripts and some other files, will fix this :)

File: Documentation/check-build.sh
Comment (errordeveloper) [reply Documentation/check-build.sh:56]: Yes, that was intentional. The aim was to catch our fatal warnings along with spelling errors. There ought to be a way to catch everything at once without running `sphinx-build` twice, but it seems like that would require calling Sphinx from a custom Python script (in my understanding that would imply extending our `conf.py`), that should also eliminate the need for grepping thing out etc. That's only a theory, and I'd rather avoid diving into that for now.

File: Documentation/check-build.sh
Comment (errordeveloper) [reply Documentation/check-build.sh:32]: For some reason the redirect with `2>&1 > warnings.txt` ceased working actually, but simple `2>` solves the issues.

The output looks like this now:

```
$ make html
tar c requirements.txt Dockerfile \
	  | docker build --tag cilium/docs-builder -
Sending build context to Docker daemon  10.24kB
Step 1/6 : FROM docker.io/library/python:3.7.5-alpine3.10
 ---> 8922d588eec6
Step 2/6 : LABEL maintainer="maintainer@cilium.io"
 ---> Using cache
 ---> f341bfd1da19
Step 3/6 : RUN apk add --no-cache --virtual --update     aspell-en     bash     ca-certificates     enchant     git     py-pip     python     sphinx-python     && true
 ---> Using cache
 ---> f2ee248a51ae
Step 4/6 : ADD ./requirements.txt /tmp/requirements.txt
 ---> Using cache
 ---> 9b21a9a640e6
Step 5/6 : RUN pip install -r /tmp/requirements.txt
 ---> Using cache
 ---> a0be0a621785
Step 6/6 : RUN install -m 0777 -d /usr/local/lib/python3.7/site-packages/versionwarning/_static/data
 ---> Using cache
 ---> 93f6ab089dec
Successfully built 93f6ab089dec
Successfully tagged cilium/docs-builder:latest
cp -r ../api _api
docker container run --rm --workdir /src/Documentation --volume /Users/ilya/Code/cilium/Documentation/..:/src --user "501:20" cilium/docs-builder ./check-build.sh html
Validating documentation (syntax, spelling)...

Please fix the following documentation warnings:
/src/Documentation/index.rst:32: WARNING: undefined label: admin_guide (if the link has no caption the label must precede a section header)

Please fix the following spelling mistakes:
* Documentation/troubleshooting.rst:9: (bdabl) 
* Documentation/troubleshooting.rst:9: (dasd) 

If the words are not misspelled, run:
Documentation/update-spelling_wordlist.sh bdabl dasd 
make: *** [html] Error 1
```

File: Documentation/.dockerignore
Comment (qmonnet) [inline Documentation/.dockerignore]: Typo: Ingore

File: Documentation/Dockerfile
Comment (qmonnet) [inline Documentation/Dockerfile]: GitHub says end-of-line is missing?

File: Documentation/Dockerfile
Comment (qmonnet) [inline Documentation/Dockerfile:9]: Is this distinct from pyenchant, which should be installed with the requirements?

File: Documentation/check-build.sh
Comment (qmonnet) [inline Documentation/check-build.sh]: No end-of-line?

File: Documentation/update-cmdrefs.sh
Comment (qmonnet) [inline Documentation/update-cmdrefs.sh]: Nor here?

File: Documentation/Makefile
Comment (qmonnet) [reply Documentation/Makefile]: Works on my setup (GNU Make 4.2.1).

**Review by qmonnet — CHANGES_REQUESTED** [review]

File: Documentation/Dockerfile
Comment (errordeveloper) [reply Documentation/Dockerfile]: Seems like a standard vscode behaviour.. let me see if there is a setting for it.

File: Documentation/Dockerfile
Comment (errordeveloper) [reply Documentation/Dockerfile:9]: Good question... I haven't questioned these dependencies, keeping them as they are for the purpose of this PR :)

File: .github/workflows/docs.yaml
Comment (joestringer) [reply .github/workflows/docs.yaml]: Usually I think they just pull from repo updates and build on the remote infrastructure, is there something specific we need to automate here?

Comment (joestringer) [reply .github/workflows/docs.yaml]: Ah, maybe we only do this for the main cilium image so if the docs base target doesn't change often then we don't need to keep building a new image.

Comment (errordeveloper) [reply .github/workflows/docs.yaml]: Yes, this one shouldn't change very often.

File: Documentation/Makefile
Comment (joestringer) [reply Documentation/Makefile:17]: Works for me, thanks!

**Comment by joestringer** [comment]
test-me-please

**Review by joestringer — CHANGES_REQUESTED** [review]
I found that this broke the build:

```
$ make
...
  GEN   Documentation/cmdref
./update-cmdrefs.sh: line 19: /src/Documentation/../cilium/cilium cmdref -d: No such file or directory
```

The `cmdref` target depends upon the compiled binaries from the tree to generate the documentation.

File: jenkinsfiles/docs.Jenkinsfile
Comment (aanm) [reply jenkinsfiles/docs.Jenkinsfile:47]: +1 we need it in the CI mostly to detect misspelled words

File: .github/workflows/docs.yaml
Comment (aanm) [inline .github/workflows/docs.yaml:14]: Will this work if external contributors push a change?

File: Documentation/check-build.sh
Comment (aanm) [reply Documentation/check-build.sh:75]: I think I've ran this somewhere manually in the past. The main issue with it was that is extremely slow since it's literally checking if the URL exists or not.

File: .github/workflows/docs.yaml
Comment (errordeveloper) [reply .github/workflows/docs.yaml:14]: I don't see a reason why not, GitHub actions that use the API are problematic with external folks, but this is just simple read-only job.

File: Documentation/check-build.sh
Comment (errordeveloper) [reply Documentation/check-build.sh:75]: Yes, and it appears to me that does those checks one-by-one in a single Python thread :)

**Comment by errordeveloper** [comment]
> I found that this broke the build:
> 
> ```
> $ make
> ...
>   GEN   Documentation/cmdref
> ./update-cmdrefs.sh: line 19: /src/Documentation/../cilium/cilium cmdref -d: No such file or directory
> ```
> 
> The `cmdref` target depends upon the compiled binaries from the tree to generate the documentation.

@joestringer yes, it does depend on the binaries, and that's exactly how it worked before. I'd expect the binaries to be there at this stage, however due to quoting issue the script was trying to call `/src/Documentation/../cilium/cilium cmdref -d` and not `/src/Documentation/../cilium/cilium`. I fixed the quotes now, would you mind trying again, or it should actually get caught by Jenkins?

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
@errordeveloper I should hope that something in our CI runs `make` ! :-)

**Comment by errordeveloper** [comment]
test-me-please

**Comment by errordeveloper** [comment]
test-me-please

**Comment by errordeveloper** [comment]
test-me-please

**Comment by errordeveloper** [comment]
test-me-please

**Comment by errordeveloper** [comment]
With @nebril's help, this should work now. Turns out our binaries are linked with glibc, so lib6-compat was needed on alpine.

**Comment by errordeveloper** [comment]
test-me-please

**Comment by errordeveloper** [comment]
It looks like we still don't get the binaries in the right place.

**Comment by errordeveloper** [comment]
The [failure on Travis](https://travis-ci.com/github/cilium/cilium/jobs/299535110#L2418) is unrelated.

```
----------------------------------------------------------------------
FAIL: node_manager_test.go:436: IPAMSuite.TestNodeManagerManyNodes
node_manager_test.go:458:
    c.Assert(testutils.WaitUntil(func() bool { return reachedAddressesNeeded(mngr, s.name, 0) }, 5*time.Second), check.IsNil)
... value *errors.errorString = &errors.errorString{s:"timeout reached while waiting for condition"} ("timeout reached while waiting for condition")
```

**Comment by errordeveloper** [comment]
test-me-please

**Comment by joestringer** [comment]
> @nebril that looks like a flake, do we know of this one?

@errordeveloper We try to file issues for any regularly seen flake (ideally any flake) so that others in the community can search for the issue, reference it, and then we can also track how often it happens and how much it's affecting development.

I searched the failure string in the issues section and this came up:
https://github.com/cilium/cilium/issues/10615

Looking at travis, it seems the failure was arm64 and I suspect it's not a coincidence, likely it's the same underlying issue.

**Comment by errordeveloper** [comment]
@joestringer looks like travis is happy now, will wait for jenkins to finish...

**Comment by errordeveloper** [comment]
test-me-please

**Comment by errordeveloper** [comment]
test-me-please

**Comment by errordeveloper** [comment]
test-me-please

**Comment by errordeveloper** [comment]
test-me-please

**Comment by errordeveloper** [comment]
Few typos to fix in the last commit message...

**Comment by errordeveloper** [comment]
test-me-please

**Comment by errordeveloper** [comment]
Looks like this is finally passing! 🎉

File: Documentation/check-cmdref.sh
Comment (joestringer) [inline Documentation/check-cmdref.sh]:
```suggestion
    echo "HINT: to fix this, run 'make -C Documentation update-cmdref ; git commit Documentation/cmdref --message \"docs: Update cmdref\"'"
```

File: Makefile
Comment (joestringer) [inline Makefile]: (not related to this specific line, but in this area)

If I currently use `make check-docs` or `make test-docs` for simple docs validation, what do you recommend I do instead? `make -C Documentation`? `make render-docs`? (I guess the latter is generally more popular anyway)

**Review by joestringer — APPROVED** [review]
I didn't do a deep review this time since I did that before, but I noticed some minor glancing nits.

Looks like this needs a rebase. With that and the typo below fixed, I think it should be good to go. Nice work!

File: Documentation/Dockerfile
Comment (pchaigno) [inline Documentation/Dockerfile:15]: Why continue if that command fails? Won't subsequent commands fail as well?

File: .github/workflows/docs.yaml
Comment (pchaigno) [inline .github/workflows/docs.yaml:3]: Will GitHub trigger this action if someone develops a Documentation change in a fork then opens a pull request? I.e., do we need to add `pull_request` here?

File: Documentation/check-cmdref.sh
Comment (pchaigno) [reply Documentation/check-cmdref.sh]: This check currently fails for me, when running `make -j$(nproc)` on a clean clone. It reports all cmdref files as missing:
```
rm -rvf cmdref/cilium*.md
docker container run --rm --workdir /src/Documentation --volume /home/paul/cilium/Documentation/..:/src --user "1000:1000" cilium/docs-builder ./check-cmdref.sh
removed 'cmdref/cilium-agent.md'
removed 'cmdref/cilium-bugtool.md'
[...]
removed 'cmdref/cilium_version.md'
docker container run --rm --workdir /src/Documentation --volume /home/paul/cilium/Documentation/..:/src --user "1000:1000" cilium/docs-builder ./update-cmdref.sh
diff --git a/Documentation/cmdref/cilium-agent.md b/Documentation/cmdref/cilium-agent.md
deleted file mode 100644
index 2729719e1..000000000
--- a/Documentation/cmdref/cilium-agent.md
+++ /dev/null
@@ -1,174 +0,0 @@
-<!-- This file was autogenerated via cilium-agent --cmdref, do not edit manually-->
-
[...]
diff --git a/Documentation/cmdref/cilium-bugtool.md b/Documentation/cmdref/cilium-bugtool.md
deleted file mode 100644
index acd702818..000000000
--- a/Documentation/cmdref/cilium-bugtool.md
+++ /dev/null
[...]
```

File: .github/workflows/docs.yaml
Comment (errordeveloper) [reply .github/workflows/docs.yaml:3]: I believe it will, the `pull_request` event in my understanding is specially about PRs as such, you can write a bot that comments on certain PRs, for example. Push is a more general one. We can test this once it's merged to be sure :)

File: Documentation/Dockerfile
Comment (errordeveloper) [reply Documentation/Dockerfile:15]: This is just a stylistic/convenience thing, it allows you to sort package alphabetically without worrying about `\`, it doesn't do anything actually. E.g. `false && true` is gonna be `false`.

File: Makefile
Comment (errordeveloper) [reply Makefile]: Yeah `make render-docs` will run the general check logic, but it also creates the Apache container, which is what it used to do also. I think `make -C Documentation` (alias for `make -C Documentation html`) will do the the checks and just write the HTML files, I avoided promoting too many new targets to the top-level makefile for the time being.

I just double-checked, and `make -C Documentation html` certainly covers what `make check-docs` as well as `make test-docs` did.

File: Documentation/check-cmdref.sh
Comment (errordeveloper) [reply Documentation/check-cmdref.sh]: @pchaigno thanks - well spotted! I am actually not sure how and whether to fix this to be honest, I am not yet equipped to run the entire build locally, and overall the Makefile is pretty complex, this could be to do with anything really. If you posted the full output, I can take a look and see what it maybe. I am not sure if we really must support `-j`, in my experience it break quite often, unless it's specifically be take care of.
Also, `-j` is most helpful when you have to compile a lot of `.c` files in `.o` files and you are invoking the compiler for each one of them, which is something you don't need to worry about with `go build`  as it runs as many gorouties as it sees fit. It's even less useful when you are invoking `docker run`. Basically you need to get dependencies really straight for `-j` to work well, we declare some dependencies, but it's also done a little naively right now. I also have serious doubts about using make dependencies with `go build`, especially since build cache was improved recently.

**Comment by errordeveloper** [comment]
test-me-please

**Comment by errordeveloper** [comment]
test-me-please

File: Documentation/Dockerfile
Comment (pchaigno) [reply Documentation/Dockerfile:15]: Oh, I see. I thought this was the same trick one needs to use with `set -e`. Thanks for the explanation!

File: Documentation/check-cmdref.sh
Comment (errordeveloper) [reply Documentation/check-cmdref.sh]: Aside from what I said about - does `make -j$(nproc)` work on master? If it did, I will have to fix it here or follow-up with another PR.

**Comment by errordeveloper** [comment]
test-me-please

File: Documentation/check-cmdref.sh
Comment (pchaigno) [reply Documentation/check-cmdref.sh]: Yes, `make -j12` works on master and saves about 32 seconds on a 53 seconds build compared to `make`. I haven't looked into what is actually impacted (maybe BPF compilation?).

I'm sending you the full output of `make -j12 &> build-output.txt` via Slack (GitHub upload seems to be temporarily failing).

Comment (errordeveloper) [reply Documentation/check-cmdref.sh]: > saves about 32 seconds on a 53 seconds build

That's not for a clean build, right?

Comment (pchaigno) [reply Documentation/check-cmdref.sh]: I think it is. At least, I ran `make clean` and `git clean -fdx` prior to those test builds.

Comment (pchaigno) [reply Documentation/check-cmdref.sh]: With this pull request, running `make` takes 10 more seconds on the first run and 10 less seconds once Docker images are prepared. Same impact on `make -j12`. I'm down to a <10s build time! :tada:

**Comment by errordeveloper** [comment]
test-me-please

File: Documentation/check-cmdref.sh
Comment (errordeveloper) [reply Documentation/check-cmdref.sh]: That's a nice side-effect indeed,  but the main goal was to actually make doc issues fail faster with GitHub Actions instead of Jenkins.

Comment (errordeveloper) [reply Documentation/check-cmdref.sh]: The `make -j` issue should now be resolved per Slack discussion.

**Review by pchaigno — APPROVED** [review]

**Pull request merged**
